### PR TITLE
changed wall distance calculation to look for nearest element, not nearest node

### PR DESCRIPTION
**Pull request by jayantmukho**
## Proposed Changes
Changed the wall distance calculation to calculate the distance to the nearest element rather than calculating the distance to the nearest node. This significantly improves the performance of turbulence models for meshes where the mesh doesn't grow perpendicular to the wall. This pull request also contains a simple improvement to the AD recording in CADTElemClass::DetermineNearestElement that significantly reduces memory usage when running the Discrete Adjoint.

This will likely break any regression test that requires a wall distance calculation, i.e. any test that has a turbulence model. Is there a easy way of changing the regression test values that doesn't require manual changing of the values?

(I mistakenly submitted the previous PR to merge with master) 

## Related Work
NA

## PR Checklist

- [X] I am submitting my contribution to the develop branch.
- [X] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [X] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.

### CODE REVIEW

**Comment by pcarruscag** [comment]
@jayantmukho thanks for this, my fsi meshes can sometimes become... creative let's say, so I may benefit from this a lot.
Just to make sure I understand this, are you now calculating distances considering the entire surface of a boundary element instead of a single point?
Also, when you instantiate the ADT you deallocate the data right after. Since the constructor takes its arguments by reference you have no guarantee that it is not, or will not at some point, store pointers to those vectors. When it does not impact the peak working set of the code, i.e. when there are no more large allocations, I prefer to leave deallocations for end of scope, or in the case of std::vector let it do its own thing.
Cheers,
Pedro

**Comment by jayantmukho** [comment]
To answer the first question, yes, the distance calculations are done based the entire surface of a boundary element. This is done using routines written by @vdweide for the higher order solver. In short, the distance to the wall of a node in the mesh is found by finding the projection of the said node on a boundary element, and the distance is calculated using that information. 

I validated this on a host of grids, some of which had sheared elements near the wall, and this procedure seemed to give the right predictions regardless of the quality of the mesh. 

Yes, from what I understand, by the end of the ComputeWall_Distance function all the memory, weather it is the vectors or the ADT itself, should be deallocated. I based this code on what Edwin was doing so he might be able to substantiate on why he deallocates the vectors immediately.

**Comment by vdweide** [comment]
@jayantmukho, the arguments to the ADT are converted to the data structures used in the ADT. This must be done, because sometimes it is necessary to gather surface grids to all ranks for an efficient search. That is why it is safe to release the memory immediately. Note that it is not necessary to do so, but this is just to avoid that an unnecessary amount of memory is allocated.

File: Common/src/geometry_structure.cpp [hunk 1]
```diff
@@ -12561,69 +12561,127 @@ void CPhysicalGeometry::Check_BoundElem_Orientation(CConfig *config) {
...
-  /*--- Compute the total number of nodes on no-slip boundaries ---*/
 
-  nVertex_SolidWall = 0;
-  for(iMarker=0; iMarker<config->GetnMarker_All(); ++iMarker) {
-    if( (config->GetMarker_All_KindBC(iMarker) == HEAT_FLUX)  ||
-       (config->GetMarker_All_KindBC(iMarker) == ISOTHERMAL) ) {
-      nVertex_SolidWall += GetnVertex(iMarker);
+    /* Check for a viscous wall. */
+    if( (config->GetMarker_All_KindBC(iMarker) == HEAT_FLUX) ||
+      (config->GetMarker_All_KindBC(iMarker) == ISOTHERMAL) ) {
```
[the remark this answers is not in the document]
Comment (jayantmukho) [reply Common/src/geometry_structure.cpp]: That's a good point, I will add that here. Just a heads up, it wasn't present in the conditional in the original ComputeWall_Distance routine.

**Comment by talbring** [comment]
Looks good to me! Thanks to @jayantmukho and @vdweide :+1:

**Comment by jayantmukho** [comment]
Can someone validate the Conjugate Heat Transfer test case? Perhaps @oleburghardt ? I included the CHT_WALL_INTERFACE in the list of markers that are used for the wall distance computation, but that was not present in the original implementation. I can run the test case, but I don't know what to validate the case against.

**Comment by vdweide** [comment]
@jayantmukho, can we merge in this one or do you prefer to wait until the data for CHT test case has been confirmed?

**Comment by jayantmukho** [comment]
I ran the CHT simulation on develop and on this branch and the solutions are, qualitatively, quite different. So I am hesitant to merge this one before we are certain that the change is correct. Both simulations converged to similar residuals, but the resultant flow solutions are not identical.

**Comment by economon** [comment]
I have run the CHT regression test to convergence in both cases, and it looks ok to me. I think it is fine to move forward with this.

**Comment by jayantmukho** [comment]
Temperature variation for the flow solution on develop:
image

Temperature variation for the flow solution on fix_wall_distance_computation
image

Difference in the temperature solution:
image

**Comment by vdweide** [comment]
@jayantmukho, is it possible to plot the wall distance for both cases as well as the difference in wall distance? Furthermore, what is the grid you used for this test case? Do you expect large differences in the wall distances between both formulations?

**Comment by vdweide** [comment]
Looks good to me.

**Pull request merged**